### PR TITLE
Fix and refactor markdown rendering

### LINKED ISSUE #32521: The newline break is missing in Issue content
### Description

The newline break is missing in Issue content ， for example 

--------------------
edit(OK):
1
2
3

preview tab (OK):
1
2
3

actual web page display as (newline break is missing) :
123
--------------------------

also see below

---------------------------
edit(OK):
```
> 1
> 2
> 3
```
preview(OK with quote block):
| 1
| 2
| 3

actual web page display as (newline break is missing) :
| 123

---------------------------

### Gitea Version

main branch

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

main branch

### Operating System

windows

### How are you running Gitea?

windows with ms sql server

### Database

SQLite

### PR DESCRIPTION
**Pull request by wxiaoguang**
Fix  #32521

Clarify the render "metas" meanings and usages

We should use clearly defined meta attributes to control the rendering behavior, instead of guessing "mode!=document" or "metas==nil"

Some existing tests are not right, for example, when rendering wiki, the commit URLs should be parsed. Related tests are also updated.

### CODE REVIEW

File: modules/markup/html_internal_test.go [hunk 1]
```diff
@@ -270,19 +277,12 @@ func TestRender_IssueIndexPattern_Document(t *testing.T) {
...
-	testRenderIssueIndexPattern(t, "#1", "#1", &RenderContext{
-		Ctx:   git.DefaultContext,
-		Metas: metas,
-	})
-	testRenderIssueIndexPattern(t, "#1312", "#1312", &RenderContext{
+	actual, err := RenderIssueTitle(&RenderContext{
 		Ctx:   git.DefaultContext,
 		Metas: metas,
-	})
-	testRenderIssueIndexPattern(t, "!1", "!1", &RenderContext{
```
Comment (lunny) [inline modules/markup/html_internal_test.go]: Why removed this test?

Comment (wxiaoguang) [reply modules/markup/html_internal_test.go]: It is wrong.

The root problem is https://github.com/go-gitea/gitea/pull/26365 made a mess and a regression #28526

Comment (wxiaoguang) [reply modules/markup/html_internal_test.go]: It should use `RenderIssueTitle` to test, and `RenderIssueTitle` doesn't use any trick to parse the issue patterns.

image

Comment (wxiaoguang) [reply modules/markup/html_internal_test.go]: Take old tests back and renamed to TestRender_IssueIndexPattern_NoShortPattern in 863170a91a08733c1e101307ba211486357a8f1e

**Review by lunny — APPROVED** [review]

**Review by Zettat123 — APPROVED** [review]

**Comment by zjjhot** [comment]
Thank you all. :-)

**Pull request merged**